controllers/route_controller.go

```diff
@@ -0,0 +1,394 @@
+func (r *RouteReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	return lattice_runtime.HandleReconcileError(r.reconcile(ctx, req))
+}
+
+func (r *RouteReconciler) reconcile(ctx context.Context, req ctrl.Request) error {
+	route, err := r.getRoute(ctx, req)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if err := r.client.Get(ctx, req.NamespacedName, route.K8sObject()); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	if !r.isRouteRelevant(ctx, route) {
+		// not relevant
+		return nil
+	}
+
+	if !route.DeletionTimestamp().IsZero() {
+		r.log.Info("Deleting")
+		r.eventRecorder.Event(route.K8sObject(), corev1.EventTypeNormal,
+			k8s.RouteEventReasonReconcile, "Deleting Reconcile")
+		if err := r.cleanupRouteResources(ctx, route); err != nil {
+			return fmt.Errorf("failed to cleanup GRPCRoute %+v: %w", route, err)
+		}
+		updateRouteListenerStatus(ctx, r.client, route)
+		r.finalizerManager.RemoveFinalizers(ctx, route.K8sObject(), routeTypeToFinalizer[r.routeType])
+
+		// TODO delete metrics
+		return nil
+	} else {
+		r.log.Info("Adding/Updating route")
+		r.eventRecorder.Event(route.K8sObject(), corev1.EventTypeNormal,
+			k8s.RouteEventReasonReconcile, "Adding/Updating Reconcile")
+		err := r.reconcileRouteResource(ctx, route)
+		// TODO add/update metrics
+		return err
+	}
+}
+
+func (r *RouteReconciler) getRoute(ctx context.Context, req ctrl.Request) (core.Route, error) {
+	switch r.routeType {
+	case HTTP:
+		httpRoute := &core.HTTPRoute{}
+		err := r.client.Get(ctx, req.NamespacedName, httpRoute.K8sObject())
+		return httpRoute, err
+	case GRPC:
+		grpcRoute := &core.GRPCRoute{}
+		err := r.client.Get(ctx, req.NamespacedName, grpcRoute.K8sObject())
+		return grpcRoute, err
+	default:
+		return nil, fmt.Errorf("unknown route type for type %s", string(r.routeType))
+	}
+}
+
+func updateRouteListenerStatus(ctx context.Context, k8sClient client.Client, route core.Route) error {
+	gw := &gateway_api_v1beta1.Gateway{}
+
+	gwNamespace := route.Namespace()
+	if route.Spec().ParentRefs()[0].Namespace != nil {
+		gwNamespace = string(*route.Spec().ParentRefs()[0].Namespace)
+	}
+	gwName := types.NamespacedName{
+		Namespace: gwNamespace,
+		// TODO assume one parent for now and point to service network
+		Name: string(route.Spec().ParentRefs()[0].Name),
+	}
+
+	if err := k8sClient.Get(ctx, gwName, gw); err != nil {
+		return fmt.Errorf("update route listener: gw not found, gw: %s, err: %w", gwName, err)
+	}
+
+	return UpdateGWListenerStatus(ctx, k8sClient, gw)
+}
+
+func (r *RouteReconciler) cleanupRouteResources(ctx context.Context, route core.Route) error {
+	_, _, err := r.buildAndDeployModel(ctx, route)
+	return err
+}
+
+func (r *RouteReconciler) isRouteRelevant(ctx context.Context, route core.Route) bool {
+	if len(route.Spec().ParentRefs()) == 0 {
+		r.log.Infof("Ignore Route which has no ParentRefs gateway %+v ", route.Spec())
```

Comment:
+v on route spec too much for log info, should be debug, also route name probably is enough for logging

Reply:
Good point, fixed next revision!